kedro/io/data_catalog.py

```diff
@@ -170,7 +187,23 @@ def __init__(
         self._data_sets = dict(data_sets or {})
         self.datasets = _FrozenDatasets(self._data_sets)
         self.layers = layers
-
+        # Keep a record of all patterns in the catalog.
+        # {dataset pattern name : dataset pattern body}
+        self.dataset_patterns = dict(dataset_patterns or {})
```

Comment:
What is the type for `dataset_patterns`? Why would we want `dataset_patterns` as a dict but `self.dataset_patterns` as dict of dict?

Reply:
It is meant to be a dict of dict. pattern name -> unresolved config stored as a dict

Reply:
Is there a reason this one is a public property? If it is needed as a public one, it should be a frozen one so no one changes it. But in my opinion, we don't need it at all or we only need the pattern names to be public.

Also could this be made sorted by default? Dictionaries in python maintain their insertion order from Python 3.6/3.7 and if not, `OrderedDict` can be used.

We shouldn't keep the same data twice, once as `self._sorted_dataset_patterns`.